src/main/java/ro/msg/learning/shop/strategy/MostAbundantStrategy.java

```diff
@@ -0,0 +1,51 @@
+package ro.msg.learning.shop.strategy;
+
+import lombok.AllArgsConstructor;
+import ro.msg.learning.shop.dto.ProductQuantityDto;
+import ro.msg.learning.shop.dto.StockDto;
+import ro.msg.learning.shop.exception.ResourceNotFoundException;
+import ro.msg.learning.shop.model.Stock;
+import ro.msg.learning.shop.repository.ProductRepository;
+import ro.msg.learning.shop.repository.StockRepository;
+import ro.msg.learning.shop.util.StockMapper;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+@AllArgsConstructor
+public class MostAbundantStrategy implements LocationStrategy {
+
+    private final StockRepository stockRepository;
+    private final ProductRepository productRepository;
+    private final StockMapper stockMapper;
+
+    @Override
+    public List<StockDto> findLocation(List<ProductQuantityDto> products) {
+
+        Map<UUID, StockDto> map = new HashMap<>();
+
+        products.forEach(p -> {
+
+            List<Stock> availableStocks = stockRepository.findByProductAndQuantity(
```

Comment:
Could you please replace findByProductAndQuantity with findByProductIdAndQuantity? In the way you implemented, it will make a request to retrieve the product first and then retrieve the stock. From the product you don't need anything else than the id, which you already have.